File: tests/.hooks/ruby/secondary/lib/alphabetically_early/sdk_hooks/custom_security_hook.rb
# typed: true
# frozen_string_literal: true

require_relative "./types"

module AlphabeticallyEarly
  module SDKHooks
    class CustomSecurityHook < AbstractSDKHook

      def before_request(hook_ctx:, request:)
        request.headers["Idempotency-Key"] = "some-key"

        case hook_ctx.operation_id
        when "customSchemeAppId"

          raise ArgumentError, "Security source is null" if hook_ctx.security_source.nil?

          custom_security = hook_ctx.security_source.call
          unless custom_security.is_a?(Models::Operations::CustomSchemeAppIdSecurity)
            raise ArgumentError, "Security source is not of type CustomSchemeAppId"
          end

          request.headers["X-Security-App-Id"] = custom_security.app_id
          request.headers["X-Security-Secret"] = custom_security.secret
        end

        request
      end

    end
  end
end

File: tests/.hooks/csharp/secondary/HoistedSecurity/Hooks/CustomSecurityHook.cs
#nullable enable
    namespace HoistedSecurity.Hooks
{
    using System;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Threading;
    using System.Threading.Tasks;
    using HoistedSecurity.Models.Operations;


    public class CustomSecurityHook : IBeforeRequestHook
    {

        public async Task<HttpRequestMessage> BeforeRequestAsync(BeforeRequestContext hookCtx, HttpRequestMessage request)
        {
            await Task.CompletedTask;

            switch (hookCtx.OperationID)
            {
                case "customSchemeAppId":
                    if (hookCtx.SecuritySource == null)
                    {
                        throw new Exception("Security source is null");
                    }

                    CustomSchemeAppIdSecurity? customSecurity = hookCtx.SecuritySource() as CustomSchemeAppIdSecurity;
                    if (customSecurity == null)
                    {
                        throw new Exception("Security source is not of type CustomSchemeAppIdSecurity");
                    }

                    request.Headers.Add("X-Security-App-Id", customSecurity.AppId);
                    request.Headers.Add("X-Security-Secret", customSecurity.Secret);
                    break;
            }

            return request;
        }
    }
}
